src/main/kotlin/no/nav/tilleggsstonader/sak/vilkår/stønadsperiode/StønadsperiodeService.kt

```diff
@@ -1,38 +1,90 @@
 package no.nav.tilleggsstonader.sak.vilkår.stønadsperiode
 
+import no.nav.tilleggsstonader.sak.behandling.BehandlingService
+import no.nav.tilleggsstonader.sak.infrastruktur.exception.feilHvis
 import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.domain.Stønadsperiode
 import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.domain.StønadsperiodeRepository
 import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.dto.StønadsperiodeDto
 import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.dto.tilSortertDto
 import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.VilkårperiodeService
 import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
 import java.util.UUID
 
 @Service
 class StønadsperiodeService(
+    private val behandlingService: BehandlingService,
     private val stønadsperiodeRepository: StønadsperiodeRepository,
     private val vilkårperiodeService: VilkårperiodeService,
 ) {
     fun hentStønadsperioder(behandlingId: UUID): List<StønadsperiodeDto> {
         return stønadsperiodeRepository.findAllByBehandlingId(behandlingId).tilSortertDto()
     }
 
+    @Transactional
     fun lagreStønadsperioder(behandlingId: UUID, stønadsperioder: List<StønadsperiodeDto>): List<StønadsperiodeDto> {
-        // TODO valider behandling er ikke låst
+        feilHvis(behandlingService.hentBehandling(behandlingId).status.behandlingErLåstForVidereRedigering()) {
+            "Kan ikke lagre stønadsperioder når behandlingen er låst"
+        }
         validerStønadsperioder(behandlingId, stønadsperioder)
-        stønadsperiodeRepository.deleteAll() // TODO skal vi finne og oppdatere de som har blitt oppdatert/slettet
-        return stønadsperiodeRepository.insertAll(
-            stønadsperioder.map {
-                Stønadsperiode(
-                    id = it.id ?: UUID.randomUUID(),
-                    behandlingId = behandlingId,
+
+        val tidligereStønadsperioder = stønadsperiodeRepository.findAllByBehandlingId(behandlingId)
+
+        slettStønadsperioder(tidligereStønadsperioder, stønadsperioder)
+        val oppdaterteStønadsperioder = oppdaterStønadsperioder(tidligereStønadsperioder, stønadsperioder)
+        val nyeStønadsperioder = nyeStønadsperioder(behandlingId, stønadsperioder, tidligereStønadsperioder)
+        return (nyeStønadsperioder + oppdaterteStønadsperioder).tilSortertDto()
+    }
+
+    private fun slettStønadsperioder(
+        tidligereStønadsperioder: List<Stønadsperiode>,
+        stønadsperioder: List<StønadsperiodeDto>,
+    ) {
+        val gjenståendePerioder = stønadsperioder.mapNotNull { it.id }.toSet()
+        val perioderTilSletting = tidligereStønadsperioder.filterNot { gjenståendePerioder.contains(it.id) }
+        stønadsperiodeRepository.deleteAllById(perioderTilSletting.map { it.id })
+    }
+
+    private fun oppdaterStønadsperioder(
```

Comment:
Kanskje `oppdaterEksisterendeStønadsperioder` (evt. tidligere) for å gjøre det tydligere at det bare er de som eksisterer man håndterer nå

Reply:
Fremgår litt gjennom at man sender med tidligere perioder og nye, men har oppdatert https://github.com/navikt/tilleggsstonader-sak/pull/163/commits/9434df4ab64fbaa9ef4c91e7f0e9a649ea6c243c